### PR TITLE
Fix GrpcChannel handle leak in AzureManaged backend

### PR DESCRIPTION
**Pull request by torosent**
# Summary
## What changed?
- Added inline channel caching in `ConfigureGrpcChannel` class using `ConcurrentDictionary<string, Lazy<GrpcChannel>>`
- Made `ConfigureGrpcChannel` implement `IDisposable` to properly dispose channels when the `ServiceProvider` is disposed
- Channels are cached per configuration key (options name + endpoint + task hub)

## Why is this change needed?
- When the Durable Task Scheduler service is unreachable, the `IConfigureNamedOptions<DurableTaskSchedulerOptions>.Configure()` method was being called multiple times (on each retry)
- Each call created a new `GrpcChannel`, which allocates internal HTTP handlers and socket connections
- The channels were **never disposed** because when `options.Channel` is pre-set (AzureManaged case), `GetCallInvoker()` returns `default` for the disposable reference
- This caused handle/resource leaks that accumulated over time, especially when the service was unavailable

## Issues / work items
- Related #N/A (fixes handle leak issue)

---

# Project checklist
- [x] Release notes are not required for the next release
  - [ ] Otherwise: Notes added to `release_notes.md`
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [x] Breaking change?
  - [x] If yes: **No breaking changes** - internal implementation change only, public API unchanged

---

# AI-assisted code disclosure (required)
## Was an AI tool used? (select one)
- [ ] No
- [x] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [ ] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot (Claude)
- AI-assisted areas/files:
  - `src/Client/AzureManaged/DurableTaskSchedulerClientExtensions.cs` (modified)
  - `src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs` (modified)
  - `test/Client/AzureManaged.Tests/DurableTaskSchedulerClientExtensionsTests.cs` (modified)
  - `test/Worker/AzureManaged.Tests/DurableTaskSchedulerWorkerExtensionsTests.cs` (modified)
- What you changed after AI output: Simplified from separate `GrpcChannelCache` class to inline `ConcurrentDictionary<string, Lazy<GrpcChannel>>` per reviewer feedback

AI verification (required if AI was used):
- [x] I understand the code and can explain it
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing
## Automated tests
- Result: Passed
  - `test/Client/AzureManaged.Tests` - 33 tests passed
  - `test/Worker/AzureManaged.Tests` - 31 tests passed
  - `test/Shared/AzureManaged.Tests` - 20 tests passed
  - **Total: 84 tests passed**

## Manual validation (only if runtime/behavior changed)
- Not performed (behavior change is internal - channels are now cached and disposed properly)

---

# Notes for reviewers
- The `ConfigureGrpcChannel` class now uses `ConcurrentDictionary<string, Lazy<GrpcChannel>>` for thread-safe channel caching
- `Lazy<GrpcChannel>` ensures thread-safe initialization without holding locks during channel creation (avoids potential deadlocks)
- Channel keys include options name, endpoint, and task hub to ensure proper isolation
- Fire-and-forget `ShutdownAsync()` is called before `Dispose()` for graceful shutdown of in-flight RPCs
- The `volatile` keyword on the `disposed` field ensures proper memory visibility when checking disposal state from multiple threads

### CODE REVIEW

**Comment by cgillum** [comment]
I asked Copilot CLI to review this with the following prompt since the currently proposed solution looks far more complicated than I was expecting.

> Please review this PR. I'm especially interested in knowing if a significantly simpler solution is possible

Here's what the response was:

```
   PR Review Summary

   Overall Assessment: The fix correctly addresses the GrpcChannel leak, but there's a significantly simpler solution available.

   The Core Problem

   When options.Channel is pre-set (AzureManaged case), GetCallInvoker() returns default for the disposable reference (line 549 in
   GrpcDurableTaskClient.cs), meaning the channel is never disposed. Combined with IConfigureNamedOptions.Configure() being called on each
   retry, this creates a resource leak.

   Simpler Alternative

   Instead of introducing a full GrpcChannelCache class with locking, disposal tracking, and 400+ lines of test code, consider this much
   simpler fix:

   Option 1: Make the channel a Lazy<GrpcChannel> in the options class

     // In ConfigureGrpcChannel class
     class ConfigureGrpcChannel(IOptionsMonitor<DurableTaskSchedulerClientOptions> schedulerOptions) :
         IConfigureNamedOptions<GrpcDurableTaskClientOptions>
     {
         readonly ConcurrentDictionary<string, Lazy<GrpcChannel>> channels = new();

         public void Configure(string? name, GrpcDurableTaskClientOptions options)
         {
             string key = name ?? Options.DefaultName;
             DurableTaskSchedulerClientOptions source = schedulerOptions.Get(key);

             options.Channel = this.channels.GetOrAdd(
                 $"{key}:{source.EndpointAddress}:{source.TaskHubName}",
                 _ => new Lazy<GrpcChannel>(source.CreateChannel)).Value;
         }
     }

   This is ~10 lines vs 170+ lines for GrpcChannelCache, and the IConfigureOptions<T> is already registered as a singleton, so the channel
   cache lives in that singleton's state.

   Option 2: Cache directly on the scheduler options instance

   Since DurableTaskSchedulerClientOptions instances are already managed via the options system, add a lazily-created channel property there:

     // In DurableTaskSchedulerClientOptions
     Lazy<GrpcChannel>? cachedChannel;

     internal GrpcChannel GetOrCreateChannel()
     {
         this.cachedChannel ??= new Lazy<GrpcChannel>(this.CreateChannel);
         return this.cachedChannel.Value;
     }

   Concerns with Current Implementation

     - Unused methods: Replace() and TryRemove() in GrpcChannelCache are never called - they add complexity without value.
     - Fire-and-forget disposal: The DisposeChannelAsync pattern with _ = Task.Run(...) can orphan disposal if the app terminates quickly.
   Consider whether synchronous disposal is acceptable.
     - Over-engineering: The cache handles scenarios (explicit replacement, removal) that don't apply to this use case.
     - Code quality issues already flagged: Nested if statements, unused variable, generic catch clause.

   Recommendation

   Go with the simpler ConcurrentDictionary<string, Lazy<GrpcChannel>> approach directly in the ConfigureGrpcChannel class. This:

     - Fixes the leak
     - Requires no new classes
     - Has minimal test surface area
     - Is easier to understand and maintain

   If you need disposal support, implement IDisposable on ConfigureGrpcChannel and register it properly with DI (though the existing code
   doesn't dispose channels either, so this may be acceptable).
```

I can try to review this more deeply tomorrow.

File: test/Client/AzureManaged.Tests/DurableTaskSchedulerClientExtensionsTests.cs [hunk 1]
```diff
@@ -280,4 +280,92 @@ public void UseDurableTaskScheduler_WithConnectionStringAndRetryOptions_ShouldCo
...
+        mockBuilder.Object.UseDurableTaskScheduler(ValidEndpoint, ValidTaskHub, credential);
+        ServiceProvider provider = services.BuildServiceProvider();
+
+        // Resolve options to trigger channel creation
+        IOptionsMonitor<GrpcDurableTaskClientOptions> optionsMonitor = provider.GetRequiredService<IOptionsMonitor<GrpcDurableTaskClientOptions>>();
+        GrpcDurableTaskClientOptions options = optionsMonitor.Get(Options.DefaultName);
+        options.Channel.Should().NotBeNull();
+
+        // Dispose the service provider - this should dispose the ConfigureGrpcChannel which disposes channels
+        provider.Dispose();
```
Comment (cgillum) [inline test/Client/AzureManaged.Tests/DurableTaskSchedulerClientExtensionsTests.cs]: I see we're disposing the provider but there's no code that actually checks to see if the channel was disposed. I don't think we can rely on this test to know whether the disposal code is working as intended.

File: src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs [hunk 2]
```diff
@@ -119,9 +134,64 @@ class ConfigureGrpcChannel(IOptionsMonitor<DurableTaskSchedulerWorkerOptions> sc
...
+            }
+
+            this.channels.Clear();
+        }
+
+        static void DisposeChannel(GrpcChannel channel)
+        {
+            // ShutdownAsync is the graceful way to close a gRPC channel.
+            // Fire-and-forget but ensure the channel is eventually disposed.
+            _ = Task.Run(async () =>
```
Comment (cgillum) [inline src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs]: Why are we doing fire-and-forget here? This seems dangerous from a reliability perspective and makes it much harder to author reliable tests.

File: src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs [hunk 3]
```diff
@@ -101,11 +104,23 @@ static void ConfigureSchedulerOptions(
     /// <summary>
     /// Configuration class that sets up gRPC channels for worker options
     /// using the provided Durable Task Scheduler options.
+    /// Channels are cached per configuration key and disposed when the service provider is disposed.
     /// </summary>
-    /// <param name="schedulerOptions">Monitor for accessing the current scheduler options configuration.</param>
-    class ConfigureGrpcChannel(IOptionsMonitor<DurableTaskSchedulerWorkerOptions> schedulerOptions) :
-        IConfigureNamedOptions<GrpcDurableTaskWorkerOptions>
+    sealed class ConfigureGrpcChannel : IConfigureNamedOptions<GrpcDurableTaskWorkerOptions>, IDisposable
```
Comment (cgillum) [inline src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs]: Are we able to use `IAsyncDisposable` instead of `IDisposable`? The gRPC channel uses an async method to shutdown so it would be better if our dispose code could call it in the proper async way.

File: src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs [hunk 4]
```diff
@@ -119,9 +134,73 @@ class ConfigureGrpcChannel(IOptionsMonitor<DurableTaskSchedulerWorkerOptions> sc
...
+            {
+                throw new ObjectDisposedException(nameof(ConfigureGrpcChannel));
+            }
+#endif
+
+            string optionsName = name ?? Options.DefaultName;
+            DurableTaskSchedulerWorkerOptions source = this.schedulerOptions.Get(optionsName);
+
+            // Create a cache key based on the options name, endpoint, and task hub.
+            // This ensures channels are reused for the same configuration
```
[the remark this answers is not in the document]
Comment (torosent) [reply src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs:150]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/microsoft/durabletask-dotnet/pull/625#discussion_r2734149559)

File: test/Worker/AzureManaged.Tests/DurableTaskSchedulerWorkerExtensionsTests.cs [hunk 5]
```diff
@@ -198,4 +199,121 @@ public void UseDurableTaskScheduler_WithNamedOptions_ShouldConfigureCorrectly()
...
+        // Arrange
+        ServiceCollection services = new ServiceCollection();
+        Mock<IDurableTaskWorkerBuilder> mockBuilder = new Mock<IDurableTaskWorkerBuilder>();
+        mockBuilder.Setup(b => b.Services).Returns(services);
+        DefaultAzureCredential credential = new DefaultAzureCredential();
+
+        // Act
+        mockBuilder.Object.UseDurableTaskScheduler(ValidEndpoint, ValidTaskHub, credential);
+        using ServiceProvider provider = services.BuildServiceProvider();
+
```
[the remark this answers is not in the document]
Comment (torosent) [reply test/Worker/AzureManaged.Tests/DurableTaskSchedulerWorkerExtensionsTests.cs:215]: @copilot open a new pull request to apply changes based on [this feedback](https://github.com/microsoft/durabletask-dotnet/pull/625#discussion_r2737260860)

File: src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs (hunk 2 above)
Comment (torosent) [reply src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs]: fixed

File: test/Client/AzureManaged.Tests/DurableTaskSchedulerClientExtensionsTests.cs (hunk 1 above)
Comment (torosent) [reply test/Client/AzureManaged.Tests/DurableTaskSchedulerClientExtensionsTests.cs]: fixed

File: src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs (hunk 3 above)
Comment (torosent) [reply src/Worker/AzureManaged/DurableTaskSchedulerWorkerExtensions.cs]: fixed

**Comment by torosent** [comment]
@copilot open a new pull request to apply changes based on the comments in [this thread](https://github.com/microsoft/durabletask-dotnet/pull/625#pullrequestreview-3717452046)

**Comment by torosent** [comment]
Closing this PR and will open a new fixed PR
